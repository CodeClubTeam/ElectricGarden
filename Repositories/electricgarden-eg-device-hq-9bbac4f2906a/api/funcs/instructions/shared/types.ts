import {
  ResultWithHttpResponse,
  EventHubResponse,
  ActionRecord,
} from "../../shared";

interface ActionRecordTableInsertResponse {
  table: ActionRecord[];
}

export interface InstructionsBindings
  extends Partial<EventHubResponse>,
    Partial<ActionRecordTableInsertResponse> {}

export interface InstructionsResponse
  extends ResultWithHttpResponse,
    InstructionsBindings {}
