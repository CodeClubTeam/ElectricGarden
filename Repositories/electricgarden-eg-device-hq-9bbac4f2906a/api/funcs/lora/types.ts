// generated from an example payload

export interface ThingParkPayload {
  DevEUI_uplink: DevEUIuplink;
}

interface DevEUIuplink {
  Time: string;
  DevEUI: string;
  payload_hex: string;
  LrrRSSI: number;
  LrrSNR: number;
}

export type PayloadElements = {
  rssi: DevEUIuplink["LrrRSSI"];
  snr: DevEUIuplink["LrrSNR"];
  timestamp: DevEUIuplink["Time"];
  deviceId: DevEUIuplink["DevEUI"];
};

export type PayloadElementsFull = PayloadElements & {
  rawData: DevEUIuplink["payload_hex"];
};

export type RegisterDeviceMapping = (
  loraDeviceId: string,
  serial: string,
) => Promise<void>;

export type GetSerialFromDeviceId = (loraDeviceId: string) => Promise<string>;
