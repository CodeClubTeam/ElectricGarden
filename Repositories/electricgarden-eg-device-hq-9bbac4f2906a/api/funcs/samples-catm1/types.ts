import { Sample } from "../shared";

export type SampleWithOptionalTimestamp = Partial<Sample>;

export type SamplesWithSerial = {
  serial: string;
  samples: SampleWithOptionalTimestamp[];
};

export type Converter = (entries: string[]) => SamplesWithSerial;
