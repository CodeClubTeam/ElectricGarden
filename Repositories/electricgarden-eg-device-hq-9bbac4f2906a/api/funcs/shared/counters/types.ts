export type EncodedCounterEntry = {
  encodingVersion: number;
  timestamp: Date;
  type: number;
  value: number;
};

export type CounterEntry = {
  timestamp: Date;
  name: string;
  value: number;
  error?: boolean;
};
