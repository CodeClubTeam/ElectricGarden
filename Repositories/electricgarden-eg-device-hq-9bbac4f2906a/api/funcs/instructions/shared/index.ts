export * from "./getDeviceInfoOrValidationError";
export * from "./types";
export * from "./validators";
