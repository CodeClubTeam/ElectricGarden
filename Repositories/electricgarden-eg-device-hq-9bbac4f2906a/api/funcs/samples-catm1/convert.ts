import { convertV1 } from "./convertV1";
import { convertV2 } from "./convertV2";
import { Converter } from "./types";

const getConverter = (entries: string[]): Converter => {
  const formatter = entries[0].split(",");
  if (formatter.length <= 2) {
    return convertV1;
  } else {
    return convertV2;
  }
};

export const convert = (payload: string) => {
  const entries = payload.split("|");
  const convert = getConverter(entries);

  return convert(entries);
};
