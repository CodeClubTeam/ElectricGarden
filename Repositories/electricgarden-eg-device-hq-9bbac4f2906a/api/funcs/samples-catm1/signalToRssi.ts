export const signalToRssi = (value: string | undefined) => {
  if (!value) {
    return undefined;
  }
  if (isNaN(parseInt(value, 10))) {
    return undefined;
  }
  const signal = Number(value);
  if (signal >= 0 && signal <= 31) {
    return -113 + signal * 2;
  } else {
    return undefined;
  }
};
