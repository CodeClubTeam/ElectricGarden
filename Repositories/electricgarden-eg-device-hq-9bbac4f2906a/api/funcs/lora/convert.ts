import { Message } from "../shared";
import { convertBootup } from "./convertBootup";
import { convertCounters } from "./convertCounters";
import { convertError } from "./convertError";
import { convertLoRaV1, convertLoRaV2 } from "./convertSamples";
import { convertSendTime } from "./convertSendTime";
import {
  GetSerialFromDeviceId,
  PayloadElements,
  PayloadElementsFull,
  RegisterDeviceMapping,
  ThingParkPayload,
} from "./types";

export const extractPayloadElements = (
  payload: ThingParkPayload,
): PayloadElementsFull => ({
  rawData: payload.DevEUI_uplink.payload_hex,
  rssi: payload.DevEUI_uplink.LrrRSSI,
  snr: payload.DevEUI_uplink.LrrSNR,
  timestamp: payload.DevEUI_uplink.Time,
  deviceId: payload.DevEUI_uplink.DevEUI,
});

enum PayloadType {
  PING = 0,
  BOOTUP = 1,
  SAMPLE = 2,
  COUNTER = 3,
  ERROR = 4,
  TIME = 6,
}

const getMessages = async (
  elements: PayloadElements,
  type: PayloadType, // replace with enum
  payload: string,
  { registerDeviceMapping, getSerialFromDeviceId }: ConvertOptions,
): Promise<Message[]> => {
  if (type === PayloadType.PING) {
    // Message contains ADR workaround ping, no messages needed
    return [];
  }
  if (type === PayloadType.BOOTUP) {
    // Message contains bootup info
    return convertBootup(elements, payload, registerDeviceMapping);
  }
  if (type === PayloadType.SAMPLE) {
    // Message contains samples
    return convertLoRaV2(elements, payload, getSerialFromDeviceId);
  }
  if (type === PayloadType.COUNTER) {
    // Message contains counters
    return convertCounters(elements, payload, getSerialFromDeviceId);
  }
  if (type === PayloadType.ERROR) {
    // Message contains errors
    return convertError(elements, payload, getSerialFromDeviceId);
  }
  if (type === PayloadType.TIME) {
    // Message contains time request
    return convertSendTime(elements, payload, registerDeviceMapping);
  }
  // No byte prefix of 1, 2 or 3
  // so assume original sample payload
  return convertLoRaV1(elements, payload, registerDeviceMapping);
};

interface ConvertOptions {
  registerDeviceMapping?: RegisterDeviceMapping;
  getSerialFromDeviceId: GetSerialFromDeviceId;
}

interface UnpackPayloadForRoute {
  type: number; // replace with enum
  payload: string;
}

export const unpackRawData = (rawData: string): UnpackPayloadForRoute => {
  const type = parseInt(rawData.substring(0, 2), 16);
  const payload = rawData.substring(2);
  // non header e.g. 9
  if (type > 9) {
    return {
      type: 9,
      payload: rawData,
    };
  }
  return {
    type,
    payload,
  };
};

export const convert = async (
  rawPayload: ThingParkPayload,
  options: ConvertOptions,
) => {
  const { rawData, ...elements } = extractPayloadElements(rawPayload);
  const { type, payload } = unpackRawData(rawData);
  return getMessages(elements, type, payload, options);
};
