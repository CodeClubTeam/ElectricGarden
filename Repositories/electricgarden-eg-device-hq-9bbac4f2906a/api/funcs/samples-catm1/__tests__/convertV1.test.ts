/* eslint prefer-const: 0 */ // --> OFF
import { Sample } from "../../shared";
import { convert } from "../convert";
import { signalToRssi } from "../signalToRssi";
import { SamplesWithSerial } from "../types";

type RawInput = {
  deviceRaw: string;
  timestampRaw: string;
  signalRaw?: string;
};

const buildPayloadFromRaw = (rawInput: RawInput[]) => {
  const payload = rawInput
    .map(({ deviceRaw, timestampRaw, signalRaw }) =>
      signalRaw
        ? `${deviceRaw};${timestampRaw};${signalRaw}`
        : `${deviceRaw};${timestampRaw}`,
    )
    .join("|");
  return payload;
};

interface DeviceData extends Omit<Sample, "timestamp" | "snr"> {
  serial: string;
}

type PropName = keyof Omit<DeviceData, "serial">;

const buildDeviceRaw = ({
  serial,
  ambientHumidity,
  ambientTemp,
  batteryVoltage,
  light,
  probeAirTemp,
  probeMoisture,
  probeSoilTemp,
  errorCode,
}: DeviceData) => {
  const values = [
    serial,
    probeSoilTemp,
    probeAirTemp,
    batteryVoltage,
    ambientTemp,
    probeMoisture,
    light,
    ambientHumidity,
  ];
  if (errorCode !== undefined) {
    values.push(errorCode ?? 0);
  }
  return values.join(":");
};

const buildTimestampRaw = (timestamp?: string) => (timestamp ? timestamp : "");

const defaultTimestampInputStr = "18/09/14,13:08:30+52";
const defaultTimestampMilliseconds = Date.parse("2018-09-14T13:08:30+13:00");
const defaultTimestamp = new Date(defaultTimestampMilliseconds);

const buildSignalRaw = (signal?: string) => (signal ? signal : "");

const defaultSignal = "20";

type Input = {
  deviceData: DeviceData;
  timestamp?: string;
  signal?: string;
};

// turns human readable form into the encoded data used for transport
const buildPayload = (inputs: Input[]) => {
  const payload = buildPayloadFromRaw(
    inputs.map(
      ({ deviceData, timestamp, signal }): RawInput => {
        const result: RawInput = {
          deviceRaw: buildDeviceRaw(deviceData),
          timestampRaw: buildTimestampRaw(timestamp),
        };
        if (signal !== undefined) {
          result.signalRaw = buildSignalRaw(signal);
        }
        return result;
      },
    ),
  );
  return payload;
};

const normalizeTimestamp = (timestamp: Date) =>
  Math.floor(timestamp.getTime() / 1000);

const serial = "ABC456";

describe("catm1 convert (v1)", () => {
  let inputs: Input[];
  let payload: any;
  let result: SamplesWithSerial;

  const runConvert = (input: Input[]) => {
    payload = buildPayload(input);
    result = convert(payload);
  };

  const getFirstInputSerial = () => {
    return inputs[0].deviceData.serial;
  };

  const getFirstInputEntry = () => {
    return inputs[0];
  };

  const getFirstOutputEntry = () => {
    return result.samples[0];
  };

  beforeEach(() => {
    inputs = [
      {
        deviceData: {
          serial: "3HLL8DL",
          probeSoilTemp: 28.5,
          probeAirTemp: 28.3,
          batteryVoltage: 1.04,
          ambientTemp: 28.3,
          probeMoisture: 1.5,
          light: 63.5,
          ambientHumidity: 33.0,
        },
        timestamp: defaultTimestampInputStr,
        signal: defaultSignal,
      },
    ];
  });

  describe.only("received single entry payload", () => {
    beforeEach(() => {
      runConvert(inputs);
    });

    it("result matches snapshot", () => {
      expect(result).toMatchSnapshot("sample");
    });

    it("serial entry exists", () => {
      expect(result.serial).toBe(getFirstInputSerial());
    });

    it("only one sample", () => {
      expect(result.samples).toHaveLength(1);
    });

    it("values match device data", () => {
      const entry = getFirstOutputEntry();
      for (const [key, value] of Object.entries(
        getFirstInputEntry().deviceData,
      )) {
        if (["serial"].includes(key)) {
          continue;
        }
        if (entry[key as PropName] !== value) {
          console.log(key);
        }
        expect(entry[key as PropName]).toBe(value);
      }
    });

    it("timestamp matches input payload", () => {
      const { timestamp } = getFirstOutputEntry();
      expect(timestamp).toBeDefined();
      if (timestamp) {
        expect(normalizeTimestamp(timestamp)).toBe(
          normalizeTimestamp(defaultTimestamp),
        );
      }
    });
  });

  describe("received multi entry payload", () => {
    beforeEach(() => {
      inputs.push({
        deviceData: {
          serial,
        },
        timestamp: defaultTimestampInputStr,
      });
      inputs.push({
        deviceData: {
          serial,
        },
        timestamp: defaultTimestampInputStr,
      });
      runConvert(inputs);
    });

    it("serial entry exists for first", () => {
      expect(result.serial).toBe(serial);
    });

    it("entries combined so three samples", () => {
      expect(result.samples).toHaveLength(3);
    });
  });

  // describe.skip('with error code', () => {
  //   beforeEach(() => {
  //     deviceData.errorCode = 123;
  //     runConvert(deviceData);
  //   });

  //   it('result matches snapshot', () => {
  //     expect(result).toMatchSnapshot('with error code');
  //   });

  //   it('error code matches input', () => {
  //     expect(getFirstOutputEntry()[getOutputKey("errorCode")]).toBe(deviceData.errorCode?.toString());
  //   });
  // });

  describe("no timestamp", () => {
    beforeEach(() => {
      inputs[0].timestamp = undefined;
      runConvert(inputs);
    });

    it("undefined should be returned on timestamp property", () => {
      expect(result.samples[0].timestamp).toBeUndefined(); // not the job of convert to filter
    });
  });

  describe("signal", () => {
    describe("value of 20", () => {
      beforeEach(() => {
        inputs[0].signal = "20";
        runConvert(inputs);
      });

      it("rssi should be output of signalConvert", () => {
        expect(getFirstOutputEntry().rssi).toBe(signalToRssi("20"));
      });
    });

    describe("empty value", () => {
      beforeEach(() => {
        inputs[0].signal = "";
        runConvert(inputs);
      });

      it("rssi should be undefined", () => {
        expect(getFirstOutputEntry().rssi).toBeUndefined();
      });
    });

    describe("undefined value", () => {
      beforeEach(() => {
        delete inputs[0].signal;
        runConvert(inputs);
      });

      it("rssi should be undefined", () => {
        expect(getFirstOutputEntry().rssi).toBeUndefined();
      });
    });
  });

  describe("Given version 2 hex payloads", () => {
    describe("valid reading", () => {
      beforeEach(() => {
        let expectedReading =
          "3HLL8F8,19/12/06,13:44:26+52,20,01|0001000000b400c8028b00c8000f01ec";
        result = convert(expectedReading);
      });

      it("An array is returned", () => {
        expect(result).toBeDefined();
      });
      it('rssi should be defined and equal "-73"', () => {
        expect(getFirstOutputEntry().rssi).toBe("-73");
      });
      it('soil temp should be valid and equal to "20.0"', () => {
        expect(getFirstOutputEntry().probeSoilTemp).toBe("20.0");
      });
      it("timestamp should be defined", () => {
        expect(getFirstOutputEntry().timestamp).toBe(1575593066);
      });
    });

    describe("invalid reading", () => {
      beforeEach(() => {
        let invalidReading =
          "3HLL8F8,19/12/06,13:44:26+52,99,01|0001000000b400c8028b00c8000f01e";
        result = convert(invalidReading);
      });

      it("An undefined output is returned", () => {
        expect(getFirstOutputEntry()).toBeUndefined();
      });
    });

    describe("missed reading (0x8000 value)", () => {
      beforeEach(() => {
        let missedReading =
          "3HLL8F8,19/12/06,13:44:26+52,20,01|0001000000b48000028b00c8000f01ec";
        result = convert(missedReading);
      });

      it("Undefined returned for amb temp", () => {
        expect(getFirstOutputEntry().ambientTemp).toBeUndefined();
      });

      it("A correct value is returned for soil temp", () => {
        expect(getFirstOutputEntry().probeSoilTemp).toBe("20.0");
      });
    });
  });
});
