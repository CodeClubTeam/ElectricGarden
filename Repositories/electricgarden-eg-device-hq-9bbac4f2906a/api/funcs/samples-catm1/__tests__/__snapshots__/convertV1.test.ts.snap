// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catm1 convert (v1) received single entry payload result matches snapshot: sample 1`] = `
Object {
  "samples": Array [
    Object {
      "ambientHumidity": 33,
      "ambientTemp": 28.3,
      "batteryVoltage": 1.04,
      "errorCode": undefined,
      "light": 63.5,
      "probeAirTemp": 28.3,
      "probeMoisture": 1.5,
      "probeSoilTemp": 28.5,
      "rssi": -73,
      "timestamp": 2018-09-14T00:08:30.000Z,
    },
  ],
  "serial": "3HLL8DL",
}
`;
